### PR TITLE
[ssr][lazyload] prevent page flicker

### PR DESCRIPTION
**Pull request by aitboudad**
left `Ng2LocationConfig` only to be fully compatible with ssr which I'll provide a fix for it in another PR

### CODE REVIEW

**Comment by christopherthielen** [comment]
I don't know `APP_INITIALIZER`, what does this do? Waits for the first transition, before the first render?

**Comment by aitboudad** [comment]
block the bootstrap until the first transition is resolved, before the render

File: src/uiRouterNgModule.ts [hunk 1]
```diff
@@ -16,12 +16,30 @@ import { ROUTES } from "@angular/router";
 /** @hidden */ export const UIROUTER_STATES       = new OpaqueToken("UIRouter States");
 // /** @hidden */ export const ROUTES = UIROUTER_STATES;
 
+export function onTransitionReady(transitionService: TransitionService, plateformId) {
+    if (isPlatformServer(plateformId)) {
+       return () => Promise.resolve();
+    }
+
+    return () => new Promise(resolve => {
```
Comment (christopherthielen) [inline src/uiRouterNgModule.ts]: could this be simplified to something like this?
```
router.globals.start$.take(1).map(trans => trans.promise.then(noop, noop)).toPromise()
```

Comment (aitboudad) [inline src/uiRouterNgModule.ts]: I got a circular deps, I'll try to find a solution!

Comment (christopherthielen) [inline src/uiRouterNgModule.ts]: How about this:
```
    if (isPlatformServer(platformId)) {
       return () => Promise.resolve();
    }

    return () => new Promise(resolve => {
      const hook = trans => trans.promise.then(done, done).then(resolve);
      const done: any = transitionService.onBefore({}, hook);
    });
```

Comment (aitboudad) [inline src/uiRouterNgModule.ts]: update into:
```ts
    return () => new Promise(resolve => {
      const hook = trans => { trans.promise.then(resolve, resolve); };
      transitionService.onBefore({}, hook);
    });
```
the circular deps is an issue on my side, `router.globals.start$` should work too but I prefer this one as it's more simple

**Comment by christopherthielen** [comment]
After thinking about this further, I'm not sure this should be the default behavior. There are plenty of apps that do not load an initial state, and instead rely on the root component and the unnamed (root) state.  Here is an example of what I am referring to: http://plnkr.co/edit/VYon4E38czpU7jSpt3Kx?p=preview

Thoughts @aitboudad ?

File: src/uiRouterNgModule.ts [hunk 2]
```diff
@@ -16,12 +16,24 @@ import { ROUTES } from "@angular/router";
...
+      transitionService.onSuccess({}, resolve);
```
Comment (aitboudad) [inline src/uiRouterNgModule.ts]: I think it should work, there is something wrong in my app that prevent it and I didn't have the time to deep into it right now

**Comment by aitboudad** [comment]
I think you're right, moving it into doc is enough I think!

**Comment by christopherthielen** [comment]
We could also add a toggle like `deferInitialRender` in the `forRoot()` (or elsewhere)

**Comment by aitboudad** [comment]
even better :+1: , I'll work on it with `Ng2LocationConfig` next week

**Comment by mkoczorowski** [comment]
is this merged ?

**Comment by aitboudad** [comment]
@mkoczorowski not yet, I'll try to find some time this week to get rid with that part

**Comment by kolkov** [comment]
Why we still use Ng2 prefix?

**Comment by aitboudad** [comment]
closed in favor of #160

**Pull request closed without merging**